staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_discovery.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
```

Comment:
I do not believe it's appropriate to go directly to v1 with this.  That is not how we would introduce a new API, and I don't think the bar is lower for meta.

Reply:
Our thought process is since this is not an API that is persisted/connected with storage, it only serves as something clients import for serialization/deserialization. 

cc @deads2k @lavalamp

Reply:
We can't make kubectl use an alpha API. Probably not a beta one either. So doing alpha -> beta -> v1 would delay all impact of this feature by 8 months, and when we got to v1 we would have no more experience at that point than we do today, because the feature will have had no users.

And if we did break the rules to use alpha right away, we'd end up with a disaster in kubectl in a few releases, where it's doing a complicated dance searching for the latest api both it and apiserver speak.

As Jeff says, there's no impact to storage.

If we want to evolve the api further (i.e. make it watchable) we will make a completely different endpoint elsewhere anyway.

The alpha -> beta -> GA progression is intended to protect us from mistakes we don't realize we are making, but in this case we already have an API as prior art. I am willing to bet that we are not making an important mistake. (As evidence, look at the comments you have made-- I don't think any of them have risen to the level where I would regret not having delayed 8 months to find them.)

An 8 month delay makes this feature not worth the effort at all IMO.

That said, v1 is inappropriate for this, there is already a v1. This should be v2.

Reply:
(Jeff pointed out that this is in the meta group -- I think we should make an apidiscovery/v2 group/version for this. @deads2k or @liggitt may have a better idea?)

Reply:
Trying to summarize the side discussions

* Historically new v2 apis have had v2alpha1 / v2beta1 / v2 progression, just like v1
* The alpha -> beta -> ga transition for resource APIs (those that either store to etcd or wrap others, like the metricsserver APIs) is intended to capture three primary phases:
    * Alpha: gather feedback with no promise of forward compatibility
    * Beta: commitment to api evolution without regression except for key feedback prior to GA
    * GA: the decision to actually commit to a no-regression / no-deprecation process
    * The distinction between alpha and beta is forward stability (we don't expose non-stable apis by default), the difference between beta and GA is the commitment to long term maintenance in the project
* Non-resource APIs generally are introduced in effectively stable form (promising forward stability), things like:
    * /healthz checks - can't be changed without breaking deployers
    * /metrics endpoints - although individual metrics may have weaker stability guarantees
    * discovery apis - used by clients we do not control and must be stable and predictable
    * remote execution endpoints such as the kubelet's log, exec, and proxy endpoints
    * These typically look more like a feature gate progression: off-by-default, on-by-default, on-forever
    * When the client is not under our control on-by-default means "API does not regress"
* The discussion about whether watch was required was a key decision point between whether discovery is a resource or non-resource API (implementing watch on non-resource would be painful)
    * At this time we believe discovery does not need inherent watch to satisfy key consumers of the new aggregated discovery
* The choice of new endpoint per resource type creates burden for the consumer as the API evolves (clients shouldn't have to make multiple HTTP requests to safely evolve over time)
   * It has been suggested that content type negotiation on the `/apis` endpoint matches the desired goal of "give me a different representation of the discovery doc without recursion" AND avoids the need over time to make multiple requests to newer versions - instead the client simply asks for the desired versions it understands in order and the server negotiates.

The clarification and consensus we are trying to reach here is:

1. Discovery is a non-resource API class
2. As a non-resource API class, once the feature gate is "on-by-default" the API is required to be stable (only additive features)
3. Non-resource APIs that are "off-by-default" do not promise stability
4. A non-resource APIs that has to change before promotion to "on-by-default" must represent incompatible changes somehow to clients (if the version is "v1" and then we find a bug, we would have to rev to "v2" before "on-by-default", which means "v1" might not ever be exposed to end users)
5. Unversioned net new endpoints (`/healthz`) are effectively v1 even if they are "off-by-default"
6. We don't want to have multiple endpoints for discovery because it's confusing for users and defeats the purpose of making discovery more efficient, and we have a way to do that with negotiation

I think 4 is important:

> Clients (like kubectl) that want to use the "off-by-default" schema need to distinguish between breaking changes to that schema

If we want to use content negotiation, kubectl needs to send *some* value to the server.  That value is implicitly a version identifier.  If the server silently changes the definition of that schema across time, then old clients are broken.  So that value is either "v<N>alpha1", implying the schema MUST change before it is GA, or the value is "v<N>beta1", implying the schema COULD change.

I would probably be ok with:

1. Feature gate off by default in release N
2. Schema version / group / kind being distinct from meta, version is "v1beta1", Kind is what is proposed here
3. kubectl in release N could request the v1beta1 type conditionally *if* we have sufficient testing that the introduction of a new type doesn't break kubectl (via test cases?)
4. In release N+1 we either rev to "v1beta2" (if we find schema issues) or go "on-by-default"
5. In release N+2 we can rev to "v1" and drop the gate

That's not "on by default in one release", but it moves this from a "this is an API that is deliberately unstable" to "an API we may evolve".  The presence of "v1beta1" in release N prevents clients from assuming "v1beta1" won't be removed, the fallback of content negotiation to the original type ensures they gracefully degrade when v1beta1 is removed in the future (which would be tested in the N release by the gate off tests).

Reply:
Added most of this discussion to the KEP. Resolving this.